flume-ng-sinks/flume-dataset-sink/src/main/java/org/apache/flume/sink/kite/DatasetSink.java

```diff
@@ -285,94 +310,222 @@ public Status process() throws EventDeliveryException {
+
+      return view.newWriter();
+    } catch (DatasetNotFoundException ex) {
+      throw new EventDeliveryException("Dataset " + datasetUri + " not found."
+          + " The dataset must be created before Flume can write to it.", ex);
+    } catch (RuntimeException ex) {
+      throw new EventDeliveryException("Error trying to open a new"
+          + " writer for dataset " + datasetUri + ": "
+          + ex.getLocalizedMessage(), ex);
     }
   }
 
-  private static Schema schema(Event event) throws EventDeliveryException {
-    Map<String, String> headers = event.getHeaders();
-    String schemaURL = headers.get(
-        DatasetSinkConstants.AVRO_SCHEMA_URL_HEADER);
+  /**
+   * Return true if the sink should roll the writer.
+   * 
+   * Currently, this is based on time since the last roll.
+   * 
+   * @return True if and only if the sink should roll the writer
+   */
+  private boolean shouldRoll() {
+    long currentTimeMillis = System.currentTimeMillis();
+    long elapsedTimeSeconds = TimeUnit.MILLISECONDS.toSeconds(
+        currentTimeMillis - lastRolledMs);
+
+    LOG.debug("Current time: {}, lastRolled: {}, diff: {} sec",
+        new Object[] {currentTimeMillis, lastRolledMs, elapsedTimeSeconds});
+
+    return elapsedTimeSeconds >= rollIntervalS;
+  }
+
+  /**
+   * Roll the dataset writer. If there's an error during the roll, the
+   * transaction will be rolled back.
+   * 
+   * @param closeWriter If true, the current writer is closed
+   * @param createNewWriter If true, a new writer is created
+   * @throws EventDeliveryException There was an error during the roll which
+   *                                means we can't guarantee previously written
+   *                                records were persisted.
+   */
+  @VisibleForTesting
+  void rollWriter(boolean closeWriter, boolean createNewWriter)
+      throws EventDeliveryException {
     try {
-      if (headers.get(DatasetSinkConstants.AVRO_SCHEMA_URL_HEADER) != null) {
-        return schemasFromURL.get(schemaURL);
-      } else {
-        return schemasFromLiteral.get(
-            headers.get(DatasetSinkConstants.AVRO_SCHEMA_LITERAL_HEADER));
+      // close the current writer and get a new one
+      if (writer != null && closeWriter) {
+        try {
+          writer.close();
+        } catch (DatasetIOException ex) {
+          throw new EventDeliveryException("Check HDFS permissions/health. IO"
+              + " error trying to close the  writer for dataset " + datasetUri
+              + ": " + ex.getLocalizedMessage(), ex);
+        } catch (DatasetWriterException ex) {
+          throw new EventDeliveryException("Failure moving temp file: "
+              + ex.getLocalizedMessage(), ex);
+        } catch (RuntimeException ex) {
+          throw new EventDeliveryException("Error trying to close the  writer for"
+              + " dataset " + datasetUri + ": " + ex.getLocalizedMessage(), ex);
+        }
+      }
+      writer = null;
```

Comment:
You don't always want to close the writer because if the writer has thrown a RuntimeException then it won't be in the Error state and doing a close will move the temp file and guarantee there is duplicated data.

This would be cleaner with CDK-885 in.